validphys2/src/validphys/pseudodata.py

```diff
@@ -236,10 +236,146 @@ def indexed_make_replica(groups_index, make_replica):
+        by Level 0 fake data.
+
+    Example
+    -------
+    >>> from validphys.api import API
+    >>> API.level0_commondata_wc(dataset_inputs = [{"dataset":"NMC"}], use_cuts="internal", theoryid=200,fakepdf = "NNPDF40_nnlo_as_01180")
+
+    [CommonData(setname='NMC', ndata=204, commondataproc='DIS_NCE', nkin=3, nsys=16)]
+    """
+    from validphys.covmats import dataset_t0_predictions
+    level0_commondata_instances_wc = []
+
+    # ==== Load validphys.coredata.CommonData instance with cuts ====#
+
+    for j, dataset in enumerate(data.datasets):
+        if dataset.cuts is None:
+            commondata_wc = dataset.commondata.load_commondata_instance()
+        else:
+            cuts = dataset.cuts.load()
+            commondata_wc = dataset.commondata.load_commondata_instance().with_cuts(
+                cuts
+            )
+
+        # == Generate a new CommonData instance with central value given by Level 0 data generated with fakepdf ==#
+
+        t0_prediction = dataset_t0_predictions(
+            dataset=dataset, t0set=fakepdf
+        )  # N.B. cuts already applied to th. pred.
+        level0_commondata_instances_wc.append(
+            commondata_wc.with_central_value(t0_prediction)
+        )
+
+    return level0_commondata_instances_wc
+
+
+def make_level1_data(
+    data, commondata_wc, level0_commondata_wc, filterseed, experiments_index
+):
+    """
+    Given a list of level0 commondata instances, return the same list
+    with central values replaced by level1 data
+
+
+    Parameters
+    ----------
+
+    data : validphys.core.DataGroupSpec
+
+    commondata_wc : list
+                    list of validphys.coredata.CommonData instances corresponding to
+                    all datasets within one experiment. Cuts already applied.
+
+    level0_commondata_wc : list
+                        list of validphys.coredata.CommonData instances corresponding to
+                        all datasets within one experiment. The central value is replaced
+                        by Level 0 fake data. Cuts already applied.
+
+    filterseed: int
+                random seed used for the generation of Level 1 data
+
+
+    Returns
+    -------
+    list
+        list of validphys.coredata.CommonData instances corresponding to
+        all datasets within one experiment. The central value is replaced
+        by Level 1 fake data.
+
+    Example
+    -------
+
+    >>> from validphys.api import API
+    >>> dataset='NMC'
+    >>> l1_cd = API.make_level1_data(dataset_inputs = [{"dataset":dataset}],use_cuts="internal", theoryid=200,
+                             fakepdf = "NNPDF40_nnlo_as_01180",filterseed=1)
+    >>> l1_cd
+    [CommonData(setname='NMC', ndata=204, commondataproc='DIS_NCE', nkin=3, nsys=16)]
+    """
+
+    # =============== generate experimental covariance matrix ===============#
+    from validphys.covmats import dataset_inputs_covmat_from_systematics
```

Comment:
Please put this import at the top (there are already other `covmats` imports so it should be ok)